dotnet/src/Connectors/VectorData.Abstractions/ConnectorSupport/VectorStoreRecordModelBuilder.cs

```diff
@@ -0,0 +1,493 @@
+            // So we ignore the config here as well.
+            // TODO: Consider throwing here instead of ignoring
+            if (definitionProperty.StoragePropertyName is not null && !this.Options.UsesExternalSerializer)
+            {
+                property.StorageName = definitionProperty.StoragePropertyName;
+            }
+
+            switch (definitionProperty)
+            {
+                case VectorStoreRecordKeyProperty definitionKeyProperty:
+                    if (property is not VectorStoreRecordKeyPropertyModel keyPropertyModel)
+                    {
+                        throw new InvalidOperationException($"Property '{property.ModelName}' is not a key property.");
+                    }
+
+                    break;
+
+                case VectorStoreRecordDataProperty definitionDataProperty:
+                    if (property is not VectorStoreRecordDataPropertyModel dataProperty)
+                    {
+                        throw new InvalidOperationException($"Property '{property.ModelName}' is not a data property.");
+                    }
+
+                    dataProperty.IsFilterable = definitionDataProperty.IsFilterable;
+                    dataProperty.IsFullTextSearchable = definitionDataProperty.IsFullTextSearchable;
+
+                    break;
+
+                case VectorStoreRecordVectorProperty definitionVectorProperty:
+                    if (property is not VectorStoreRecordVectorPropertyModel vectorProperty)
+                    {
+                        throw new InvalidOperationException($"Property '{property.ModelName}' is not a vector property.");
+                    }
+
+                    if (definitionVectorProperty.Dimensions is not null)
+                    {
+                        vectorProperty.Dimensions = definitionVectorProperty.Dimensions;
+                    }
+
+                    if (definitionVectorProperty.IndexKind is not null)
+                    {
+                        vectorProperty.IndexKind = definitionVectorProperty.IndexKind;
+                    }
+
+                    if (definitionVectorProperty.DistanceFunction is not null)
+                    {
+                        vectorProperty.DistanceFunction = definitionVectorProperty.DistanceFunction;
+                    }
+
+                    break;
+
+                default:
+                    throw new ArgumentException($"Unknown type '{definitionProperty.GetType().FullName}' in vector store record definition.");
+            }
+        }
+    }
+
+    /// <summary>
+    /// Extension hook for connectors to be able to customize the model.
+    /// </summary>
+    protected virtual void Customize()
+    {
+    }
+
+    /// <summary>
+    /// Finalizes the model after all properties have been processed, validating the model and applying some final transformations.
+    /// </summary>
+    protected virtual void Finalize(Type clrType)
```

Comment:
Finalization has a certain meaning in .NET (related to finalizers aka destructors).

Some ideas that come to my mind: `Finish`, `Seal`.

It's not a blocker for merging the PR, just something to keep in mind.